### PR TITLE
add sniff sockopts

### PR DESCRIPTION
**Pull request by dd-dent**
WIP as `_socket.fromfd` doesn't work properly.  
Also, shouldn't work on mac...

Fixes: #251

### CODE REVIEW

**Comment by dd-dent** [comment]
Just so it's clear, I haven't yet abandoned this, just incredibly busy with work and RL...  
Will probably have time to work on it this weekend.

**Comment by njsmith** [comment]
No worries, work and RL definitely come first :-)

File: trio/tests/test_socket.py [hunk 1]
```diff
@@ -259,6 +259,35 @@ class MySocket(stdlib_socket.socket):
         assert s.family == tsocket.AF_INET6
 
 
+async def test_sniff_sockopts():
+    # iterate over the combinations of families/types we're testing:
+    for family_type in ['ipv4_tcp', 'ipv4_udp', 'ipv6_tcp', 'ipv6_udp']:
+        family, type = family_type.split('_')
+        family = stdlib_socket.AF_INET if family == 'ipv4' else stdlib_socket.AF_INET6
+        type = stdlib_socket.SOCK_STREAM if type == 'tcp' else stdlib_socket.SOCK_DGRAM
```
[the remark this answers is not in the document]
Comment (dd-dent) [reply trio/tests/test_socket.py]: Changed it, did a bit of wiggling to avoid over indentation/yapf butchering my code...

**Comment by dd-dent** [comment]
RE newsfragments, is this considered a feature? Seems like a bit of a stretch, and I'm pretty sure this isn't a bugfix...  
So I'm thinking misc?

**Comment by njsmith** [comment]
(FYI: I edited the initial comment to include a `Fixes: #251` note. This does two things: makes it easy to get from this to the issue and vice-versa, and it makes it so that if this PR is merged, then github will automatically mark that issue as closed, which is something we tend to forget to do otherwise :-).)

I dug into this a bit more on Windows. Apparently the answer is that I was [working from incorrect information](https://bugs.python.org/issue28134#msg332704), and it isn't expected to work on Windows! Also AFAICT it can't work on Windows – [even the `getsockname` hack doesn't work on Windows](https://bugs.python.org/issue27377#msg270669). So we should definitely skip the test on Windows. I'm glad we tried though, because otherwise we would have kept walking around with the wrong idea about how this worked! Validating assumptions is good.

Speaking of validating assumptions, it looks like the macOS tests did fail exactly the way we expected. So... that's nice to have confirmed, I guess? And, squinting at this output:

```
>           assert tsocket_socket.family == socket.family
E           AssertionError: assert <AddressFamily.AF_INET: 2> == <AddressFamily.AF_INET6: 30>
E            +  where <AddressFamily.AF_INET: 2> = <trio.socket.socket fd=27, family=AddressFamily.AF_INET, type=SocketKind.SOCK_DGRAM, proto=0, laddr=('::e82b:b90e:100:0', 0, 0, 1)>.family
E            +  and   <AddressFamily.AF_INET6: 30> = <socket.socket fd=27, family=AddressFamily.AF_INET6, type=SocketKind.SOCK_DGRAM, proto=0, laddr=('::', 0, 0, 0)>.family
```

I notice that the `tsocket_socket`'s repr includes `laddr=('::e82b:b90e:100:0', 0, 0, 1)`, which is an `AF_INET6` address, even though the socket object has `family=AF_INET`. This suggests to me that Python's `getsockname` does detect the family correctly, even if it doesn't match the `family` field in the socket object. Which suggests that if we do implement the `getsockname` hack, then it will work on macOS.

OTOH AFAICT there's no hope of getting `type` autodetection working on macOS. Oh well.

Up to you whether you want to skip the test on macOS and leave it at that for now, or implement the `getsockname` hack and test just `family` on macOS.

> RE newsfragments, is this considered a feature? Seems like a bit of a stretch, and I'm pretty sure this isn't a bugfix...

I think calling it a feature is fine. Before you had to specify manually, now you don't, yay. It's not the *biggest* feature, but who's counting?

**Comment by njsmith** [comment]
I finally had a chance to dig into this a bit tonight.

It's extremely weird.

Running with 3.7 on windows, I do always get a failure on that test. The error happens when `_sniff_sockopts_for_fileno` tries to make its temporary socket object, by doing `sockobj = _stdlib_socket.socket(fileno=fileno)` – for some reason, on py37 on Windows only, this fails consistently in this test. BUT, it works if I try simpler things at the REPL, like `fd = socket.socket().detach(); socket.socket(fileno=fd)`. Super weird!

I guess there is a workaround: make that line be

```python
sockobj = _stdlib_socket.socket(family, type, proto, fileno=fileno)
```

(I.e., passing in the "initial guesses" that `_sniff_sockopts_for_fileno` receives.)

I think there's a Python bug here that we really ought to isolate and report though...

**Comment by njsmith** [comment]
@tiran I know you made some changes to the behavior of `socket.socket(fileno=fileno)` in 3.7. This is a shot in the dark, but have you had any reports of strange failures on Windows? (See previous comment in this thread.)

**Comment by tiran** [comment]
I remember an issue with ``proto``. It was always ``0`` on Windows.

**Comment by dd-dent** [comment]
@njsmith A couple of notes:  
This issue appeared following a rebase, but I'm uncertain if the rebase had anything to do with this (IIRC, the PR was broken in some other way prior to rebasing).  
RE making this work, It's possible to possibly bypass the issue with a `_sys.platform == "linux"` check prior to calling `_sniff_sockopts_for_fileno`, although I hadn't encountered it anywhere else in trio, so I'll assume it's considered a bad practice?

**Comment by tiran** [comment]
Aaah, there is another issue. On most platforms it is not possible to retrieve socket information from a new, unconfigured socket fd. The socket most either be connected to a peer or in listening mode. Linux is one of the few platforms that supports this for the majority of sockets.

**Comment by dd-dent** [comment]
@tiran So, by extension, one might say this feature is only applicable to linux?  
Also, it's implied from what you're saying no all socket can be 'sniffed', so how do I go about finding which subset of sockets are 'sniffable'?

**Comment by tiran** [comment]
The behavior depends on the platform and platform version.

* For connected sockets it should work on all platforms.
* For unconnected sockets, you can only try and catch the error. I'm not aware of any API.

One more thing, AFAIK Windows doesn't have the concept of protocols. In my experiments, the protocol value was always 0.

**Comment by njsmith** [comment]
@tiran Sorry, my question is much more specific :-). 

As far as I can tell, calling `socket` with a `fileno` and no other arguments used to always return a socket object (possibly with some messed up metadata):

```python
socket.socket(fileno=fileno)  # where 'fileno' is a valid socket descriptor
```

But starting in 3.7, on Windows only, it now *sometimes* raises this exception instead:

```
OSError: [WinError 10022] An invalid argument was supplied: 'family'
```

We don't yet understand what the condition is to trigger this; I can't reproduce it at the REPL. But our test suite triggers it 100% of the time, so there is something consistent about it :-).

The question was whether you have any thoughts on:

* Where this error comes from
* Whether it should be considered a regression in 3.7 (since this didn't error out in previous versions AFAICT)
* Why this might be happening in our test suite but not the REPL

**Comment by dd-dent** [comment]
The relevant log section from appveyor:
```python
================================== FAILURES ===================================
_________________________ test_custom_socket_factory __________________________
    async def test_custom_socket_factory():
        class CustomSocketFactory:
            def socket(self, family, type, proto):
                return ("hi", family, type, proto)
    
        csf = CustomSocketFactory()
    
        assert tsocket.set_custom_socket_factory(csf) is None
    
        assert tsocket.socket() == ("hi", tsocket.AF_INET, tsocket.SOCK_STREAM, 0)
        assert tsocket.socket(1, 2, 3) == ("hi", 1, 2, 3)
    
        # socket with fileno= doesn't call our custom method
        fd = stdlib_socket.socket().detach()
>       wrapped = tsocket.socket(fileno=fd)
..\trio\tests\test_socket.py:840: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\trio\_socket.py:294: in socket
    family, type, proto, fileno
..\trio\_socket.py:307: in _sniff_sockopts_for_fileno
    sockobj = _stdlib_socket.socket(fileno=fileno)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
self = <socket.socket fd=-1, family=AddressFamily.AF_UNSPEC, type=0, proto=0>
family = -1, type = -1, proto = -1, fileno = 1324
    def __init__(self, family=-1, type=-1, proto=-1, fileno=None):
        # For user code address family and type values are IntEnum members, but
        # for the underlying _socket.socket they're just integers. The
        # constructor of _socket.socket converts the given argument to an
        # integer automatically.
        if fileno is None:
            if family == -1:
                family = AF_INET
            if type == -1:
                type = SOCK_STREAM
            if proto == -1:
                proto = 0
>       _socket.socket.__init__(self, family, type, proto, fileno)
E       OSError: [WinError 10022] An invalid argument was supplied: 'family'
C:\Python37\lib\socket.py:151: OSError
````  
It seems like it has something to do with the cpython's `_socket.socket` init on windows, called originally with only an fd.  
I'm now pushing a commit which both:  

1. Only runs the relevant test on linux.  
2. Only executes the `_sniff_sockopts_for_fileno` func on linux.  

Hopefully will work, but somewhat ugly in my opinion...

**Comment by dd-dent** [comment]
@njsmith Sorry to pester you, as I'm sure there's more important stuff than this, but is there anything new about this perchance?

**Comment by njsmith** [comment]
Thanks for pestering :-).

The explicit check for linux is a bit ugly yeah :-/. It's possible there's no alternative. I still have no idea about this failure... I tried a bit more to track it down, and it seems to be coming directly from some Windows API, not anything inside Python.

What do you think about the idea upthread, of changing it so `_sniff_sockopts_for_fileno` does

```python
sockobj = _stdlib_socket.socket(family, type, proto, fileno=fileno)
```

i.e., passes in whatever family/type/proto the user gave us? Rationale:

1. In my tests this seemed to get the tests passing
2. If we're on a system where we can't usefully sniff anything, then we're ultimately going to end up making that exact call anyway, later. So at least it isn't making anything worse...

**Comment by njsmith** [comment]
Looks like this is now passing everywhere *except* Linux? Frustrating :-(

**Comment by njsmith** [comment]
Oof, I'm sorry this was so painful! I guess there must be some nicer way to write it, but we've run into so many weird issues here that I'm not even sure how to find it... so maybe we should stick with this version that seems to work :-). I did edit your newsfragment; you can take a look and see if you like the new version. And thanks for your patience with this so-called "easy first issue"...

**Comment by dd-dent** [comment]
No need to feel sorry about anything, it's just what it is (bloody annoying at times).
Your version of the newsfragment is far superior to mine, so let's go with that one!

**Comment by njsmith** [comment]
Okay, merged! :tada: :birthday: And, no pressure, but if you'd like to keep contributing then we'd love to have you, so I'm sending you a github invite now. You can read more about this in [our contributing documentation](https://trio.readthedocs.io/en/latest/contributing.html#joining-the-team).

**Pull request merged**